lib/Doctrine/ORM/Persisters/ManyToManyPersister.php

```diff
@@ -328,267 +193,438 @@ public function removeElement(PersistentCollection $coll, $element)
+     */
+    protected function getDeleteSQLParameters(PersistentCollection $coll)
+    {
+        $mapping    = $coll->getMapping();
+        $identifier = $this->uow->getEntityIdentifier($coll->getOwner());
+
+        // Optimization for single column identifier
+        if (count($mapping['relationToSourceKeyColumns']) === 1) {
+            return array(reset($identifier));
+        }
+
+        // Composite identifier
+        $sourceClass = $this->em->getClassMetadata($mapping['sourceEntity']);
+        $params      = array();
+
+        foreach ($mapping['relationToSourceKeyColumns'] as $columnName => $refColumnName) {
+            $params[] = isset($sourceClass->fieldNames[$refColumnName])
+                ? $identifier[$sourceClass->fieldNames[$refColumnName]]
+                : $identifier[$sourceClass->getFieldForColumn($columnName)];
+        }
+
+        return $params;
+    }
+
+    /**
+     * Gets the SQL statement used for deleting a row from the collection.
      *
-     * @return string The SQL query part to add to a query.
+     * @param \Doctrine\ORM\PersistentCollection $coll
+     *
+     * @return string
      */
-    public function getFilterSql($mapping)
+    protected function getDeleteRowSQL(PersistentCollection $coll)
     {
-        $targetClass = $this->em->getClassMetadata($mapping['targetEntity']);
-        $rootClass   = $this->em->getClassMetadata($targetClass->rootEntityName);
-        $filterSql   = $this->generateFilterConditionSQL($rootClass, 'te');
+        $mapping = $coll->getMapping();
+        $class   = $this->em->getClassMetadata($mapping['sourceEntity']);
+        $columns = array();
 
-        if ('' === $filterSql) {
-            return array('', '');
+        foreach ($mapping['joinTable']['joinColumns'] as $joinColumn) {
+            $columns[] = $this->quoteStrategy->getJoinColumnName($joinColumn, $class, $this->platform);
         }
 
-        // A join is needed if there is filtering on the target entity
-        $tableName    = $this->quoteStrategy->getTableName($rootClass, $this->platform);
-        $joinSql      = ' JOIN ' . $tableName . ' te' . ' ON';
-        $onConditions = $this->getOnConditionSQL($mapping);
+        foreach ($mapping['joinTable']['inverseJoinColumns'] as $joinColumn) {
+            $columns[] = $this->quoteStrategy->getJoinColumnName($joinColumn, $class, $this->platform);
+        }
 
-        $joinSql .= implode(' AND ', $onConditions);
+        return 'DELETE FROM ' . $this->quoteStrategy->getJoinTableName($mapping, $class, $this->platform)
+            . ' WHERE ' . implode(' = ? AND ', $columns) . ' = ?';
+    }
 
-        return array($joinSql, $filterSql);
+    /**
+     * Gets the SQL parameters for the corresponding SQL statement to delete the given
+     * element from the given collection.
+     *
+     * @internal Order of the parameters must be the same as the order of the columns in getDeleteRowSql.
```

Comment:
This should not be documented this way. It is a usage of ``@internal`` which does not match the phpdoc standards. ``@internal`` is about marking a method as being internal, not about providing advanced documentation which should be hidden while other parts are visible